### DISCUSSION TITLE
Differences and purposes of Tinput, Thot, refTempK, and refDens

### DISCUSSION THREAD

**Discussion by drewj-usnctech** [terrapower/armi#694] [Q&A]
So far as I can tell, `Tinput` and `Thot` help determine expansion of Components. `Tinput` is the temperature where `Component` dimensions are given, but `Thot` is the temperature modeled. So the physical models would reflect the geometry after expanding components from `Tinput` to `Thot`. 

I get more confused when we throw `refTempK` in the mix. It's only defined for some materials as the "reference temperature." https://github.com/terrapower/armi/search?q=refTempK shows hastelloy and inconel alloys, SiC, and Zr. Hastelloy uses it for `linearExpansionPercent`

https://github.com/terrapower/armi/blob/b709d3450f6b7a89376abed3a2bf9107bcfe39f2/armi/materials/hastelloyN.py#L133-L151

Inonel allows use it in `polyfitLinearExpansionPercent` which doesn't appear to be used

https://github.com/terrapower/armi/blob/b709d3450f6b7a89376abed3a2bf9107bcfe39f2/armi/materials/inconel600.py#L140-L181

but I imagine there's a need for it in a mechanics plugin.

`refDens` is defined for just about everything and used heavily. Not a lot of confusion on what it means, but more on how it relates to `refTempK`. 

My original assumption would be that `refTempK` is the temperature at which `refDens` is defined. As such, would I be correct in assuming the density of a `Component` at a temperature of `refTempK` is `refDens`? This is the case for `SiC` at least, or very close to correct

```python
>>> from armi.reactor.components.basicShapes import Circle
>>> from armi.materials.siC import SiC
>>> s = SiC()
>>> s.p.refTempK
298.15
>>> s.p.refDens
3.21
>>> circ = Circle("demo", s, Tinput=s.p.refTempK-273.15, Thot=s.p.refTempK-273.15, od=1)
>>> circ.density()
3.1597241245892773
>>> circ.density() / s.p.refDens - 1
-0.015662266483091214
```
About 1.5% difference so not a massive difference, but not nothing

**Comment by drewj-usnctech** [comment]
@john-science @ntouran @keckler re-upping for some insight

**Reply by john-science** [reply]
Okay, so if you do a `git grep refTempK` in ARMI, you get:

```
armi/materials/hastelloyN.py:        self.p.refTempK = 273.15 + 20
armi/materials/hastelloyN.py:        refTempC = getTc(Tk=self.p.refTempK)

armi/materials/inconel600.py:        self.p.refTempK = 294.15
armi/materials/inconel600.py:        refTempC = getTc(None, Tk=self.p.refTempK)

armi/materials/inconel625.py:        self.p.refTempK = 294.15
armi/materials/inconel625.py:        refTempC = getTc(None, Tk=self.p.refTempK)

armi/materials/inconel800.py:        self.p.refTempK = 273.15 + 21.0
armi/materials/inconel800.py:        refTempC = getTc(Tk=self.p.refTempK)

armi/materials/inconelX750.py:        self.p.refTempK = 294.15
armi/materials/inconelX750.py:        refTempC = getTc(None, Tk=self.p.refTempK)

armi/materials/materialParameters.py:        pb.defParam("refTempK", units="K", description="Reference temperature")

armi/materials/siC.py:        self.p.refTempK = 298.15
```

It appears that `refTempK` is just the temperature upon which we decide the thermal linear expansion is zero.  So, somewhere between 0C and room temp.

I am not sure I even like that this parameter exists. It is not used in enough places I think it really needs to exist.  Maybe I will remove it entirely.

**Reply by john-science** [reply]
@ntouran  Could you help us grok the difference between `Material.density()` and `Material.p.refDens`?  It _feels_ like there is some internal inconsistency here, probably historical, which is making it a bit hard for us to disentangle what they are supposed to be used for.

For instance, it might be that they have very clear purposes, but aren't used correctly in `siC`.

**Comment by keckler** [comment]
Your understanding of `Thot` and `Tinput` are exactly correct. But I think the meaning of `refTempK` is a little more ad hoc. 

In the case of `SiC` that you point out, you can actually see that its `density()` method overwrites that of the base class and defines some `rho0` and `Tc0` of its own:
https://github.com/terrapower/armi/blob/20ca6877c2e2221bc3e3521cf9e827cf2d4a4376/armi/materials/siC.py#L99-L106

I also just noticed that, based on the last line in that method, the density is being thermally-expanded in 3D for `SiC`. This is a bug and is in direct contradiction to what the base `density()` method indicates should be happening:
https://github.com/terrapower/armi/blob/20ca6877c2e2221bc3e3521cf9e827cf2d4a4376/armi/materials/material.py#L337-L370

So both of these facts could explain why in your demonstration above that the density from `circ.density()` does not align exactly with that of `s.p.refDens`. 

What is the meaning of this? I don't really know, unfortunately. To me, it indicates that the construction of the `SiC` class is not internally-consistent, and thus it has issues. Clearly it has not been tested as thoroughly as it should be.

From my experience, many of the material classes in the public-facing ARMI framework are lacking in a variety of ways, whether that be lack of properties, inconsistent properties, or just flat-out bugs. This is especially the case for (what we at TP have historically considered) more esoteric materials like `SiC`. 

So I understand that this does not really answer your question of "what is `refTempK`?" To my knowledge, `refTempK` is not used consistently. I think it probably is supposed to be used in conjunction with `refDens` as you indicate, but obviously you've found a violation of that principle, and it is not done consistently throughout all material classes. 

All this goes to say, my opinion is that the material classes in the framework need to be carefully vetted before being used for "real" analysis.

Hopefully someone like @ntouran or @jakehader might have a better answer for you.
